### DISCUSSION TITLE
OSM maps to Lanelet

### DISCUSSION THREAD

**Discussion by ramdort** [CommonRoad/commonroad-scenario-designer#20] [Q&A] [answered]
I am working on building a Lanelet maps of a city. Common road allows this conversions from osm to osm(lanelet). 
I have done small sections of the city, and was really help. However, In a part of the city during the conversion from OSM to commonroad and then from commonroad to OSM(lanelets) I got a lot of logs and the generated file was not readble by JOSM.
![image](https://github.com/user-attachments/assets/261401ad-a327-4660-8c59-9a0578730a5e)
the below figure shows part of the log
![image](https://github.com/user-attachments/assets/c8455471-38e7-49c2-9732-2a7c108ac518)

**Comment by smaierhofer** [comment]
If you provide the source OSM map, we can try to solve the problem for the next release of the scenario designer

**Reply by ramdort** [reply]
I will provide more context to what I am trying to achieve. 
I want to create a Lanelet map of a complete town (Central business district ). 
I have tired to do it for the complete town and did not work. (same the above message)  
Then I have tired to do it in portions It shows a good performance except for some a part of the town. However, I wanted after that to merge all Lanelets  together in a one Lanelet through merge function in JOSM, but  the Lanelets are then messed up and connected to each other (no longer usable nor readable). 
The map OSM file is [here](https://file.io/fqWJxPm93G3G)
If you are aware of any tool or reference for generating Lanelets from OSM maps would be nice if you share it here. 

Regards

**Reply by smaierhofer** [reply, marked as answer]
I fixed the bug which resulted to the problem with the provided map.
The fix is pushed to the develop branch.

We support two approaches:     
- OSM->CommonRoad->Lanelet2 (as you are doing) 
- SUMO/OSM->OpenDRIVE->CommonRoad->Lanelet2 (can also be completely executed by APIs provided within this repository but you need to install SUMO manually).

**Comment by ramdort** [comment]
Thank you so much for the quick support.
Now the files are not corrupted open-able with JOSM however the logs when converting it  from osm to common road still persist. like invalid lanlets , ERROR - Fallback GeonamesID used and other warnings as figures show. 
![image](https://github.com/user-attachments/assets/83f28dd8-9bae-4037-90c4-aab6c159f64c)
![image](https://github.com/user-attachments/assets/af7db00b-cd18-465b-b3f3-0afb11d41207)
![image](https://github.com/user-attachments/assets/cee5e7a1-cb2f-49dd-bfd7-0e7d919676ad)
I have also inspected the result osm file It performed really good in most of the intersections however some parts got wrong not surprising. 
![image](https://github.com/user-attachments/assets/46d81506-f4e8-43b6-b63b-21935ae58af7)
How do you recommend dealing with it ? 

thanks

**Reply by ramdort** [reply]
I want to add one more point. 
I have used the generated lanelet2 file  from  commonroad to create a graph out of it using the lanelet2 library as [here](https://github.com/fzi-forschungszentrum-informatik/Lanelet2/blob/master/lanelet2_routing/README.md) obviously it failed. terminate called after throwing an instance of 'lanelet::NoSuchPrimitiveError'
  what():  Failed to lookup element with id 4984315. 
To put in one sentence. It successfully converts it to lanelet2(osm)  but it is not compatible with the lanelet2 library and its functionality. Do you have any suggestion to make my pipeline works 
OSM -> commonroad->lanelet2-> graph (.gv) file?

**Reply by smaierhofer** [reply]
- The warnings are mainly features/elements of OpenStreetMap, we do not support yet. 
- The wrong part of the intersection might be the result of an unusual/wrong OSM layout (since it is user created). 
We have a tool for map verification/repairing (also part of the scenario designer), but I'm not sure whether it works for such a complex intersection. 
- without knowing the details of the lanelet2 error and what concrete map element causes the error, I cannot help to fix the pipeline (I assume it is cause by some conversion accuracy from OSM->CommonRoad since the data of OSM is usually incomplete)

**Reply by ramdort** [reply]
I have converted the map successfully from osm -> commonroad -> lanelet2 -> graph however it works fine for a map in germany. It did not work for a map in US ( section of  the famous manhatten map) not big deal for me. However I have an important observation, the relationships between lanelets does not have the relationship type (left and right). my maps and the sample provided with common road scenario designer show that. It could be an issue in commonroadscenario designer.

**Reply by smaierhofer** [reply]
For the US, you probably need to update the coordinate projection. 
Regarding your observation: Do you mean left and right adjacent lanelet relationships?

**Comment by ramdort** [comment]
I will check how to update the coordinate projection. 
there are several possible relationships between lanelets as the following code shows 
```
enum class RelationType : uint8_t {
  None = 0,                 //!< No relation
  Successor = 0b1,          //!< A (the only) direct, reachable successor. Not merging and not diverging.
  Left = 0b10,              //!< (the only) directly adjacent, reachable left neighbour
  Right = 0b100,            //!< (the only) directly adjacent, reachable right neighbour
  AdjacentLeft = 0b1000,    //!< directly adjacent, unreachable left neighbor
  AdjacentRight = 0b10000,  //!< directly adjacent, unreachable right neighbor
  Conflicting = 0b100000,   //!< Unreachable but with overlapping shape
  Area = 0b1000000          //!< Adjacent to a reachable area
};
```
My observation is that none of lanelets in the provided sample and my maps that  got any ( Left or Right ) relationships not the case of the sample map in lanelet repo.
